x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,341 @@
+    private volatile long lastRequestedSeqno;
+    private volatile long globalCheckpoint;
 
-    ShardFollowNodeTask(long id, String type, String action, String description, TaskId parentTask, Map<String, String> headers) {
+    private volatile int numConcurrentReads = 0;
+    private volatile int numConcurrentWrites = 0;
+    private volatile long processedGlobalCheckpoint = 0;
+    private volatile long currentIndexMetadataVersion = 0;
+    private final AtomicInteger retryCounter = new AtomicInteger(0);
+    private final Queue<Translog.Operation> buffer = new PriorityQueue<>(Comparator.comparing(Translog.Operation::seqNo).reversed());
+
+    ShardFollowNodeTask(long id, String type, String action, String description, TaskId parentTask, Map<String, String> headers,
+                        ShardFollowTask params, BiConsumer<TimeValue, Runnable> scheduler, TimeValue idleShardChangesRequestDelay,
+                        TimeValue retryTimeout) {
         super(id, type, action, description, parentTask, headers);
+        this.params = params;
+        this.scheduler = scheduler;
+        this.retryTimeout = retryTimeout;
+        this.idleShardChangesRequestDelay = idleShardChangesRequestDelay;
+    }
+
+    void start(long followGlobalCheckpoint) {
+        this.lastRequestedSeqno = followGlobalCheckpoint;
+        this.processedGlobalCheckpoint = followGlobalCheckpoint;
+        this.globalCheckpoint = followGlobalCheckpoint;
+
+        // Forcefully updates follower mapping, this gets us the leader imd version and
+        // makes sure that leader and follower mapping are identical.
+        updateMapping(imdVersion -> {
+            currentIndexMetadataVersion = imdVersion;
+            LOGGER.info("{} Started to follow leader shard {}, followGlobalCheckPoint={}, indexMetaDataVersion={}",
+                params.getFollowShardId(), params.getLeaderShardId(), followGlobalCheckpoint, imdVersion);
+            coordinateReads();
+        });
     }
 
+    private synchronized void coordinateReads() {
+        if (isStopped()) {
+            LOGGER.info("{} shard follow task has been stopped", params.getFollowShardId());
+            return;
+        }
+
+        LOGGER.trace("{} coordinate reads, lastRequestedSeqno={}, globalCheckpoint={}",
+            params.getFollowShardId(), lastRequestedSeqno, globalCheckpoint);
+        final int maxReadSize = params.getMaxBatchOperationCount();
+        while (hasReadBudget() && lastRequestedSeqno < globalCheckpoint) {
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] read [{}/{}]", params.getFollowShardId(), numConcurrentReads, from, maxReadSize);
+            long maxRequiredSeqno = Math.min(globalCheckpoint, from + maxReadSize);
+            sendShardChangesRequest(from, maxReadSize, maxRequiredSeqno);
+            lastRequestedSeqno = maxRequiredSeqno;
+        }
+
+        if (numConcurrentReads == 0) {
```

Comment:
I just relized this has a bug - we need to check if we have budget here too (add this to the list of tests :)) - we may have no readers because the write buffer is full.

Reply:
In that the case the buffer is full and we don't start a read then the shard follow task comes to a halt, because when the write buffer has been consumed nothing else will happen as this was the last point a read will be started.

Reply:
so maybe when handeling bulk shard operation responses we should also check whether a shard changes request should be fired off?

Reply:
> while in this case we only make a copy if maxBatchSizeInBytes limit has been reached.

he - I'm pretty I saw we did in some previous iterations. And yes - it must allow reads because we block reads when we don't consume the write buffer. In a sense the right moment is when you coordinate writes and you consumed some buffer elements. Both are fine with me